packages/dcos-sdk/src/NavigationService.ts

```diff
@@ -0,0 +1,80 @@
+import { injectable, inject, named } from "inversify";
+import { EventEmitter } from "events";
+import { IExtensionProvider, ExtensionProvider } from "plugin-baby";
+
+const NAVIGATION_CHANGE = Symbol("NAVIGATION_CHANGE");
+
+interface INavigationElement {
+  children?: INavigationElement[],
+  parent?: string,
+  path?: string
+}
+
+export const NavigationServiceExtension = Symbol("NavigationServiceExtension");
+
+export interface INavigationExtension {
+  getElements(): INavigationElement[];
+}
+
+function unflatten(array: INavigationElement[], parent: INavigationElement = {}) {
+  const children = array.filter((child: INavigationElement) =>
+    child.parent === parent.path
+  );
+
+  if (children.length !== 0) {
+    if (parent.path) {
+      parent.children = children;
```

Comment:
I would love us move in a direction where we don't mutate so much anymore, could we refactor this?

Reply:
We could as a separate issue. That's a winning algo I copied from Stack overflow to not spend time on that.

Reply:
hmm, okay. Not entirely happy with that but I can understand your pragmatism there 👍